test/UserDbSpec.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TypeOperators              #-}
+
+module UserDbSpec where
+
+import           Test.Hspec
+import           Test.QuickCheck
+
+import           Control.Monad.Except      (runExceptT)
+import           Control.Monad.Reader      (runReaderT)
+
+import           Servant
+import           Database.Persist.Types      (Filter)
+import           Database.Persist.Sql        (ConnectionPool, transactionUndo)
+import           Database.Persist.Postgresql (Entity (..), fromSqlKey, insert,
+                                              selectFirst, selectList, deleteWhere,
+                                              (==.), runSqlPool)
+
+import           Config (App (..), Config (..), Environment (..), makePool)
+import           Models
+import           Api.User
+
+
+runAppToIO :: Config -> App a -> IO a
+runAppToIO config app = do
+  result <- runExceptT $ runReaderT (runApp app) config
+  case result of
+    Left err -> error $ show err
+    Right a  -> return a
+
+setupTeardown :: (Config -> IO a) -> IO ()
+setupTeardown runTestsWith = do
+  pool <- makePool Test
+  migrateDb pool
+  runTestsWith $ Config { getPool = pool, getEnv = Test }
+  cleanDb pool
+  where
+    migrateDb :: ConnectionPool -> IO ()
+    migrateDb pool = runSqlPool doMigrations pool
+
+    cleanDb :: ConnectionPool -> IO ()
+    cleanDb = deleteAllUsers
+
+    deleteAllUsers :: ConnectionPool -> IO ()
+    deleteAllUsers pool = do
+      flip runSqlPool pool $ do
```

Comment:
a fave trick of mine is that infix operators don't need the `$` for `do`, so you can also write


````haskell
pool `runSqlPool` do
    deleteWhere ([] :: [Filter User])
```

if you're feeling frisky